src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs

```diff
@@ -717,19 +717,51 @@ public bool IsNext(ReadOnlySpan<T> next, bool advancePast = false)
         [MethodImpl(MethodImplOptions.NoInlining)]
         private unsafe bool IsNextSlow(ReadOnlySpan<T> next, bool advancePast)
         {
-            // Call PeekSlow directly since we know there isn't enough space.
-            Debug.Assert(UnreadSpan.Length < next.Length);
+            ReadOnlySpan<T> currentSpan = UnreadSpan;
 
-            T* t = stackalloc T[next.Length];
-            ReadOnlySpan<T> peek = PeekSlow(new Span<T>(t, next.Length));
+            // We should only come in here if we need more data than we have in our current span
+            Debug.Assert(currentSpan.Length < next.Length);
 
-            if (next.SequenceEqual(peek))
+            int length = next.Length;
+            SequencePosition nextPosition = _nextPosition;
+
+            while (true)
             {
-                if (advancePast)
+                if (!next.StartsWith(currentSpan))
                 {
-                    Advance(next.Length);
+                    break;
+                }
+
+                if (next.Length == currentSpan.Length)
+                {
+                    // Fully matched
+                    if (advancePast)
+                    {
+                        Advance(length);
```

Comment:
This Advance will traverse the sequence again, no? Can we avoid that by advancing when we get the next segment?

Reply:
We can't advance the cursor until we know we've got a full match. We could try and slice off of `nextPosition`, but it would be really convoluted and I suspect it wouldn't save much. I want to be careful about making this code too complicated for small gains.